### PR TITLE
infoschema, executor: Add SQL Text to INFORMATION_SCHEMA.(CLUSTER_)DEADLOCKS table

### PR DESCRIPTION
**Pull request by MyonKeminta**
### What problem does this PR solve?

Problem Summary: This PR adds a column to the (CLUSTER_)DEADLOCKS table that shows the current blocked SQL statements.

### What is changed and how it works?

What's Changed: This PR adds a column to the (CLUSTER_)DEADLOCKS table that shows the current blocked SQL statements.

### Check List 

Tests 

- [x] Unit test

Side effects

- [x] Performance regression: Consumes more CPU
- [x] Performance regression: Consumes more Memory

Documentation

- [ ] Affects user behaviors

### Release note 

```release-note
Add SQL text information to `INFORMATION_SCHEMA.DEADLOCKS` and `INFORMATION_SCHEMA.CLUSTER_DEADLOCKS` table
```

### CODE REVIEW

File: executor/infoschema_reader.go [hunk 1]
```diff
@@ -2350,6 +2324,143 @@ func (r *dataLockWaitsTableRetriever) retrieve(ctx context.Context, sctx session
...
+		waitChainIdx = 0
+		idx++
+		for idx < len(r.deadlocks) && len(r.deadlocks[idx].WaitChain) == 0 {
+			idx++
+		}
+	}
+	return idx, waitChainIdx
+}
+
+func (r *deadlocksTableRetriever) retrieve(ctx context.Context, sctx sessionctx.Context) ([][]types.Datum, error) {
```
Comment (cfzjywxk) [inline executor/infoschema_reader.go:2352]: Is it possible to abstract the retrieve implementation together with the one used by `TiDB_TRX` table?

Comment (MyonKeminta) [reply executor/infoschema_reader.go:2352]: It's kind of hard... Can we just continue this PR, and I submit another PR if I find some way to abstract it?

File: executor/infoschema_reader.go [hunk 2]
```diff
@@ -2350,6 +2324,143 @@ func (r *dataLockWaitsTableRetriever) retrieve(ctx context.Context, sctx session
...
+			if c.Name.O == deadlockhistory.ColCurrentSQLDigestTextStr {
+				if sqlRetriever == nil {
+					sqlRetriever = NewSQLDigestTextRetriever()
+				}
+
+				idx, waitChainIdx := r.currentIdx, r.currentWaitChainIdx
+				for i := start; i < end; i++ {
+					sqlRetriever.SQLDigestsMap[r.deadlocks[idx].WaitChain[waitChainIdx].SQLDigest] = ""
+					// Step to the next entry
+					idx, waitChainIdx = r.nextIndexPair(idx, waitChainIdx)
```
Comment (youjiali1995) [inline executor/infoschema_reader.go:2400]: Can it cause index-out-of-range?  `nextIndexPair` can return a `idx` which equals `len(r.deadlocks)`.

**Review by youjiali1995 — APPROVED** [review]
LGTM

File: executor/infoschema_reader.go (hunk 2 above)
Comment (MyonKeminta) [reply executor/infoschema_reader.go:2400]: It should be broken by the `for` loop condition. But maybe I can add an additional check here for safety.

Comment (youjiali1995) [reply executor/infoschema_reader.go:2408]: As long as you have confidence in safety, needn't add the check. 😂

**Review by cfzjywxk — APPROVED** [review]
/lgtm

**Pull request merged**